### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by tateviktome**
Add unit tests

### CODE REVIEW

File: RealFakeStore/RealFakeStoreTests/UnitTests/CartReducerTests.swift [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        XCTAssertEqual(next.cart, expectedCart)
+        
+        next = Reducers.cartReducer(next, .loadProducts(.failure(ProductServiceError.invalidURL)))
+        XCTAssertEqual(next.cart, expectedCart)
+        
+        next = Reducers.cartReducer(next, .loadProducts(.pending))
+        XCTAssertEqual(next.cart, expectedCart)
+        
+        next = Reducers.cartReducer(next, .loadProducts(.loading))
+        XCTAssertEqual(next.cart, expectedCart)
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/CartReducerTests.swift:62]: Looks good!
Can we also test:
- adding a product that already exists in the cart
- removing a product that does not exist in the cart
- setting quantity to 0 or negative numbers
- setting the quantity of a product that is not in the cart

File: RealFakeStore/RealFakeStoreTests/UnitTests/CartTests.swift [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+class CartTests: XCTestCase {
+    func testCartHappy() {
+        let product = Product.fromJSON().first!
+        let cart = Cart(cartProducts: [CartProduct(product: product, quantity: 10)])
+        
+        let subtotal = cart.subtotal
+        let totalCost = cart.totalCost
+        XCTAssertNotNil(Double(subtotal))
+        XCTAssertNotNil(Double(totalCost))
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/CartTests.swift:20]: Can we also check against the actual expected values for these? Idea is to make sure that the algorithm for calculating the total is correct.

File: RealFakeStore/RealFakeStoreTests/UnitTests/CartTests.swift [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        XCTAssertEqual(emptyCartTotalCost, "0.0")
+        XCTAssertFalse(emptyCart.qualifiesForFreeShipping())
+        XCTAssertNil(emptyCart.productWith(id: product.id))
+        XCTAssertFalse(emptyCart.hasProductWith(id: product.id))
+        XCTAssertNotEqual(emptyCart.quantityOf(product: product), 10)
+    }
+    
+    func testCartSad() {
+        let product = Product.fromJSON().first!
+        let cart = Cart(cartProducts: [CartProduct(product: product, quantity: -10)]) //invalid input
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/CartTests.swift:40]: Nice

File: RealFakeStore/RealFakeStoreTests/UnitTests/ProductTests.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+@testable import RealFakeStore
+
+class ProductTests: XCTestCase {
+    func testProductHappy() {
+        XCTAssertNoThrow(Product.fromJSON())
+    }
+    
+    func testProductSad() {
+        
+    }
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/ProductTests.swift:19]: I think we don't really have a sad path for this (or do we?)

File: RealFakeStore/RealFakeStoreTests/UnitTests/RandomTests.swift [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+import Foundation
+@testable import RealFakeStore
+import XCTest
+
+class RandomTests: XCTestCase {
+    func testArrayUpdateHappy() {
+        var array = Product.fromJSON()
+        
+        let expectedArray = array.map { product -> Product in
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/RandomTests.swift:16]: Nice!

File: RealFakeStore/RealFakeStoreTests/UnitTests/RandomTests.swift [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+                           image: product.image,
+                           description: product.description,
+                           price: 0.0)
+        }
+        
+        XCTAssertEqual(array, expectedArray)
+    }
+    
+    func testArrayUpdateSad() {
+        //update an element that doesnt exist
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/RandomTests.swift:41]: We don't really have a sad path here. Perhaps an empty array?

File: RealFakeStore/RealFakeStoreTests/UnitTests/ResolvableTests.swift [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        XCTAssertNotNil(failureResolvable.error)
+    }
+    
+    func testResolvableSad() {
+        let pendingResolvable = Resolvable<Product>.pending
+        let loadingResolvable = Resolvable<Product>.loading
+        let successResolvable = Resolvable.success(Product.fromJSON())
+        let failureResolvable = Resolvable<Product>.failure(ProductServiceError.invalidURL)
+        
+        XCTAssertFalse(pendingResolvable.isLoading)
```
Comment (WickedLynx) [inline RealFakeStore/RealFakeStoreTests/UnitTests/ResolvableTests.swift:32]: Nice!

**Review by WickedLynx — APPROVED** [review]
Looks good! Just need to add a few more asserts for the cart reducer

File: RealFakeStore/RealFakeStoreTests/UnitTests/CartReducerTests.swift (hunk 1 above)
Comment (tateviktome) [reply RealFakeStore/RealFakeStoreTests/UnitTests/CartReducerTests.swift:62]: This cases were covered already in lines 43-50
- adding a product that already exists in the cart
- removing a product that does not exist in the cart
- setting quantity to 0 or negative numbers

Just added
- setting the quantity of a product that is not in the cart

File: RealFakeStore/RealFakeStoreTests/UnitTests/ProductTests.swift (hunk 4 above)
Comment (tateviktome) [reply RealFakeStore/RealFakeStoreTests/UnitTests/ProductTests.swift:19]: I just changed some things about this
So I added `static func fromJSONOrEmpty()` functional to replace our  `static func fromJSON()` in method calls.
The thing was that `fromJSON()` never threw the errors, just returned empty array if errors occurred, but if someone actually messes the JSON file a bit, we'll never see the test fail.

So, I made `fromJSON()` to throw errors, and `fromJSONOrEmpty` to return `fromJSON()` or empty for testing purposes.

File: RealFakeStore/RealFakeStore/Source/Store/Actions/Actions.swift [hunk 8]
```diff
@@ -16,11 +16,11 @@ enum ActionType {
 }
 
 protocol Action {
-    func exec(dispatcher: Dispather)
+    func exec(dispatcher: Dispatcher)
```
Comment (cavoco) [inline RealFakeStore/RealFakeStore/Source/Store/Actions/Actions.swift:19]: Just had to rename this, it was bothering me :)

**Comment by cavoco** [comment]
I am wondering if it is a good idea to add a `Testing` target and exclude the previews from these tests? (To up the code coverage numbers)

File: RealFakeStore/RealFakeStore/Source/Model/Product.swift [hunk 9]
```diff
@@ -17,7 +17,7 @@ struct Product: Decodable {
 }
 
 extension Product {
-    static func fromJSON() -> [Product] {
+    static func fromJSON() throws -> [Product] {
         guard let fileURL = Bundle.main.url(forResource: "products", withExtension: "json") else {
             return []
         }
```
Comment (cavoco) [inline RealFakeStore/RealFakeStore/Source/Model/Product.swift:23]: Would it be useful to test how the app handles the resource name being incorrect or if we try to decode an invalid json file?

**Review by cavoco — APPROVED** [review]

File: RealFakeStore/RealFakeStore/Source/Model/Product.swift (hunk 9 above)
Comment (WickedLynx) [reply RealFakeStore/RealFakeStore/Source/Model/Product.swift:23]: Yeah, we had a chat about it. If it is invalid the function now throws. We assert that it doesn't throw.
So the test is to ensure that the file is not invalid and the decoding logic is correct.

**Pull request merged**